man/fi_mr.3.md

```diff
@@ -736,6 +743,9 @@ The follow flag may be specified to any memory registration call.
   API AllocHost call (as opposed to using malloc-like host allocation,
   unified/shared memory allocation, or AllocDevice).
 
+*FI_HMEM_DATA_GDRCOPY_HANDLE*
+: This flag indicates that hmem_data points to a gdrcopy_handle struct.
```

Comment:
struct gdrcopy_handle is something internal to libfabric.  A user of the API won't have this.  For the external API, this should be *gdr_mh_t.  If additional information beyond gdr_mh_t is needed, then we will need to define a structure in the exported API files for callers to use.  Such a structure cannot create a dependency between the libfabric build and cuda (or any other GPU library).

Reply:
We could actually expose gdrcopy_handle for the application to do the memory pinning etc. and pass it to libfabric.
```
// src/hmem_cuda_gdrcopy.c
struct gdrcopy_handle {
	gdr_mh_t mh; /* memory handler */
	void *cuda_ptr; /* page aligned gpu pointer */
	void *user_ptr; /* user space ptr mapped to GPU memory */
	size_t length; /* page aligned length */
};
```

Because of gdrcopy's page alignment requirements, `gdr_mh_t` alone is not sufficient for address translation, so we do need to define a different struct for callers to use.

libfabric build should not have a hard dependency on gdrcopy, and we should skip gdrcopy if it is not available.

Reply:
We cannot have gdr_mh_t in any of the libfabric external header files (include/rdma/*.h) without forcing a build dependency on the gdrcopy library, which is a no-go.  The register mr APIs already have parameters to pass in an array of addresses and lengths.  To register this memory for use with gdr copy, which of the ptrs do we actually need?